client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -232,13 +232,14 @@ String refreshToken() {
         } else {
             log.debug("Token is already under refresh for scope {} and stream {}", this.scopeName, this.streamName);
         }
-        try {
-            currentRefreshFuture.join(); // Block until the token is refreshed
-        } finally {
-            this.tokenRefreshFuture.compareAndSet(currentRefreshFuture, null); // Token is already refreshed, so resetting the future to null.
-        }
-        LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
-        return delegationToken.get().getValue();
+
+        final CompletableFuture<Void> handleToCurrentRefreshFuture  = currentRefreshFuture;
+        return currentRefreshFuture.thenApply(v -> {
+                    // Token is already refreshed, so resetting the future to null.
+                    this.tokenRefreshFuture.compareAndSet(handleToCurrentRefreshFuture, null);
```

Comment:
What happens if `compareAndSet` returns false? Your update will not have succeeded.

Reply:
Yes, that is as intended. We want to update it to null only if it is still the same object.